### ISSUE TITLE
[Feature Request] More guidance and support for theming and restyling of components

### ISSUE THREAD

**Issue by nborelli** [ionic-team/stencil#917]
**Stencil version:**

```
 @stencil/core@0.10.2<version>
```

**I'm submitting a:**

[ ] bug report
[x] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or https://stencil-worldwide.slack.com

**Current behavior:**

It is unclear from the documentation how to support different themes for a component.  A user can learn a lot from reading the source of the Ionic components, but they shouldn't have to.  It is unclear whether we should be setting theme in the host attribute of the component decorator, adding a mode property to our components, etc.  Should the theme.ts file in the Ionic component library be part of Stencil.js or should developers be left to figure out theming on their own?

**Expected behavior:**

It would be nice to have clear documentation on how to support themes and restyling of components in both the Shadow DOM and no Shadow DOM use cases.

In addition, it would be nice to have a migration plan for the draft W3C CSS Shadow Parts specification at https://drafts.csswg.org/css-shadow-parts.  Monica Dinculescu at Google also has a blog post discussing the specification (at least the early draft) that can be found at https://meowni.ca/posts/part-theme-explainer/.  

This may be purely a task of improving the documentation, but as it stands now, it feels like Stencil is missing support for easily managing component themes in both the shadow DOM and none shadow DOM cases.  I know we can use CSS variables (except in IE11), but this breaks down pretty quickly for more complex theming.

**Comment by nborelli** [comment]
I should add that the idea of injecting the `Mode` into the component might have to be expanded beyond just 'ios' and 'md'.  We may be using Stencil to build a PWA for the desktop and/or want to create our own theming.  The relationship of `mode` to the `styleUrls` attribute of the `@Component` decorator or `ModeStyles` interface is not clear and as far as I know, we do not have control over how mode is set on our components.

Please forgive me if I am missing something obvious though.

Mode is defined as follows in Ionic:
```ts
export type Mode = 'ios' | 'md';
```

**Comment by adamdbradley** [comment]
Yes we need to improve our docs. But one thing to clear up is that stencil has is no limitation to `mode`, and can be set to whatever values you'd like. In Ionic's case, it decided to use `ios` and `md`, but stencil itself doesn't no restrict it to only those.

**Comment by nborelli** [comment]
Hi @adamdbradley,

Thanks for the reply. That is really good to know.  But how does one go about specifying the CoreContext.mode value?  I am new to this and don't understand when to do this.  I can see the global context `window.App.Context`, and could assign the mode property in the app component, but this seems wrong.  Is there a better place for us to specify the mode we want to run in?  Is this something that we do by writing a global script and adding it to the configuration.  An example would help.

Thanks,
Neal

**Comment by arjunyel** [comment]
Also stencil has polyfils so CSS Variables works in IE11 :)
